Прикрой меня тестом

GoveRaven opened the pull request.

GoveRaven committed:
дописал логику для слайс юзера и добавил тесты для него и селектора

GoveRaven committed:
добавил поле isSimilarFilmsLoading

GoveRaven committed:
изменил типизацию селекторам

GoveRaven committed:
добавил тесты для слайса с данными

GoveRaven committed:
поменял способ экспорта и импорта browserHistory

GoveRaven committed:
добаивл тесты для middleware

GoveRaven committed:
добавил функции для создания моковых данных и добавил их в тесты

GoveRaven committed:
переименовал переменную

GoveRaven committed:
добавил тесты для асинхронных операций

Lexeria commented on project/src/components/history-routes/browser-history.ts:
```diff
@@ -1,3 +1,5 @@
-export const browserHistory = createBrowserHistory();
+const browserHistory = createBrowserHistory();
+
+export default browserHistory;
```
Лучше не использовать экспорт по умолчанию, давай вернем именованный экспорт? Это нужно, чтобы везде мы использовали именно то имя, которое указываем в этом файле. И в разных местах одна и та же сущность называлась одинаково. Если нужно переименовать в каком-то другом месте, то можно сделать так:
```js
import { browserHistory as otherName } from '../....';
```

Lexeria replied:
Я увидел, что это сделано для создания мока в Jest, в котором указывается лишь путь до модуля. В таком случае могу попробовать посоветовать использовать одновременно именнованный и не именнованный экспорт:
```js
export const browserHistory = createBrowserHistory();
export default browserHistory;
```

Как минимум именованный экспорт еще удобен тем, что твой редактор кода или IDE автоматически будут предлагать и добавлять строчку с импортом выше, когда ты используешь указанное в модуле имя. Так что советую все же от него не отказываться и просто добавить экспорт по умолчанию для Jest

GoveRaven replied:
изменил

Lexeria commented on project/src/pages/movie.tsx:
```diff
@@ -14,6 +14,7 @@ import { NotFound } from './not-found';
 import {
   selectChosenFilm,
   selectIsFilmLoading,
+  selectisSimilarFilmsLoading,
```
`is` должно начинать с большой буквы, чтобы соблюсти нейминг в стиле `camelCase`

GoveRaven replied:
исправил

Lexeria commented on project/src/pages/movie.tsx:
```diff
@@ -37,7 +39,7 @@ export function MoviePage(): JSX.Element {
   }, [dispatch, numberId]);
 
-  if (isFilmLoading) {
+  if (isFilmLoading && isSimilarFilmsLoading) {
```
А тут точно должно быть `&&`, а не `||`?

GoveRaven replied:
Я думаю, что да. Если и фильм, и похожие фильмы не загрузились, то показывать лаудер. Зачем убирать лаудер, если один из двух элементов ещё не отображается на странице?

Lexeria replied:
В том-то и дело, что один из двух, а не одновременно два. Тут нужно логическое ИЛИ, а не И

GoveRaven replied:
Да. До меня дошла моя ошибка. Я не так понял эти переменные. Исправил

Lexeria commented on project/src/services/middleware.test.ts:
```diff
@@ -0,0 +1,36 @@
+  it('should redirect to "/login" with redirectRouter action', () => {
+    store.dispatch(redirectToRoute(AppRoute.SignIn));
+    expect(fakeHistory.location.pathname).toBe(AppRoute.SignIn);
+    expect(store.getActions()).toEqual([redirectToRoute(AppRoute.SignIn)]);
```
Не обязательно, но можно вытащить вызов `redirectToRoute(AppRoute.SignIn)` в переменную выше, чтобы не дублировать

GoveRaven replied:
сделал

Lexeria commented on project/src/services/middleware.test.ts:
```diff
@@ -0,0 +1,36 @@
+    expect(store.getActions()).toEqual([redirectToRoute(AppRoute.SignIn)]);
+  });
+
+  it('should not to be redirect "/login" because bad acction', () => {
```
Просто придирка, что после некоторых модальных глаголов не ставятся инфинитивы, а значит и частичка to. Плюс еще есть парочка ошибок других в формировании предложения, поэтому оставляю это замечание просто как напоминание и рекомендация писать описание на привычном русском и не париться. Тут никто не запрещает это делать, мы ж на русском проекте. Бывают исключения, но они обычно вызваны договоренностями в команде

Lexeria commented on project/src/store/slices/data-slice/data-slice.test.ts:
```diff
@@ -0,0 +1,137 @@
+import { dataSlice } from './data-slice';
+import { createMockFilm } from '../../../utils/createMockFilm';
+import { createMockComments } from '../../../utils/createMockComments';
+describe('data slice', () => {
```
Советую оставлять отступы между импортом и началом кода. Обычно есть хотя бы одна, иногда две строчки между ними

GoveRaven replied:
в VSCode визуально кажется, что отступ есть, поэтому я упустил его
![image](https://github.com/user-attachments/assets/0f30bfb9-6e15-4ffb-9143-03ee08e2db9c)
Отступ добавил

Lexeria commented on project/src/store/slices/data-slice/selector.test.ts:
```diff
@@ -0,0 +1,113 @@
+import { datatype } from 'faker';
```
А понял, что за библиотека такая `faker` и как и зачем применяется `datatype`?

GoveRaven replied:
`faker` применяется чтобы создавать фэйковые данные. 
`datatype` нужен, чтобы генерировать типы данных

Lexeria commented on project/src/store/slices/data-slice/selector.ts:
```diff
@@ -3,35 +3,49 @@ import { TFilm } from '../../../types/films';
-export const selectPromoFilm = (state: TState): TFilm | null =>
-  state[SliceName.Data].promoFilm;
+export const selectPromoFilm = (
+  state: Pick<TState, SliceName.Data>
```
Это тип большой и везде дублируется, можно просто повыше вытащить его в `TDataState` или `TDataSliceState`

GoveRaven replied:
вынес в тип

Lexeria commented on project/src/store/slices/user-slice/selector.test.ts:
```diff
@@ -0,0 +1,83 @@
+      user: null,
+    };
+  });
+  it('should return authorization status from store', () => {
```
Мелочь, но лучше отступ для читаемости добавить

GoveRaven replied:
добавил

Lexeria commented on project/src/store/slices/user-slice/selector.ts:
```diff
@@ -3,11 +3,13 @@ import { SliceName } from '../../../consts/store-action';
-export const selectAuthStatus = (state: TState): AuthorizationStatus =>
-  state[SliceName.User].authStatus;
+export const selectAuthStatus = (
+  state: Pick<TState, SliceName.User>
```
Тоже замечание про вынесение типа повыше, что везде `Pick` не дублировать

GoveRaven replied:
добавил

Lexeria commented on project/src/store/slices/user-slice/user-slice.ts:
```diff
@@ -28,13 +27,16 @@ export const userSlice = createSlice({
       .addCase(authLogin.fulfilled, (state, action) => {
         state.authStatus = AuthorizationStatus.Auth;
         state.user = action.payload;
+        state.isAuthStatusChecked = true;
       })
       .addCase(authLogin.rejected, (state) => {
         state.authStatus = AuthorizationStatus.NoAuth;
+        state.isAuthStatusChecked = true;
       })
       .addCase(authLogout.fulfilled, (state) => {
         state.authStatus = AuthorizationStatus.NoAuth;
         state.user = null;
+        state.isAuthStatusChecked = true;
```
А мы точно должны везде это проставлять? У нас же `checkAuth` вызывается при старте приложения, а значит так или иначе мы поменяем статус после того, как действие выполнится

GoveRaven replied:
Я написал это из-за тестов. Так выходило, что после этих операций по тестам в стейте `isAuthStatusChecked  === UNKNOW`. И меня это напрягало. Хотел услышать твоё мнение

Lexeria replied:
Менять рабочий код под тесты в таком ключе не очень хорошо, лучше менять тесты с учетом этого: например, не проверять `isAuthStatusChecked`, либо добавить в `beforeEach` или в тест-кейсы действие `checkAuth`, чтобы было по-честному, как происходит в приложении

GoveRaven replied:
переписал тесты и слайс

Lexeria commented on project/src/types/store.ts:
```diff
@@ -26,6 +26,7 @@ export type TDataSlice = {
   chosenFilm: TFilm | null;
   isFilmLoading: boolean;
   similarFilms: TFilm[];
+  isSimilarFilmsLoading: boolean;
```
Правильнее `are` в начале, так как множественное число

GoveRaven replied:
заменил

Lexeria commented on project/src/utils/createMockComments.ts:
```diff
@@ -0,0 +1,15 @@
+import { TReview } from '../types/reviews';
+import { datatype, date, internet, random } from 'faker';
+
+export function createMockComments(): TReview {
```
Можно эту и другие функции для создания моков вытащить в отдельный файл один, чтобы по разным файлам не размазывать, так как у нас не так много моков создается, причем часть ты создаешь прямо в тестах без вынесения в утилзы

GoveRaven replied:
вытащил в один файл

Lexeria commented on project/src/store/slices/user-slice/selector.test.ts:
```diff
@@ -0,0 +1,83 @@
+    });
+
+    it('should return null after logout', () => {
+      const currentUser: TUser = {
+        avatarUrl: internet.avatar(),
+        email: internet.email(),
+        id: Math.round(Math.random() * (100 - 1) + 1),
+        name: internet.userName(),
+        token: internet.ip(),
+      };
```
А у нас есть утилза для этого, или есть различия какие-то специфичные, чтобы ее тут не использовать?

GoveRaven replied:
идея с утилзами пришла мне не сразу, поэтому в какой-то момент я переписывал код. Этот остался тут из-за того, что я не доглядел
Добавил утилз

Lexeria commented on project/src/store/slices/user-slice/selector.test.ts:
```diff
@@ -0,0 +1,83 @@
+    });
+
+    it('should return object', () => {
+      const currentUser: TUser = {
```
Тот же вопрос про утилзу

GoveRaven replied:
тоже добавил

Lexeria commented on project/src/pages/movie.tsx:
```diff
@@ -14,7 +14,7 @@ import { NotFound } from './not-found';
   selectChosenFilm,
   selectIsFilmLoading,
-  selectisSimilarFilmsLoading,
+  selectareSimilarFilmsLoading,
```
Переименовал на `are`, а замечание про `camelCase` пропустилось из-за этого, видимо

GoveRaven replied:
исправил

Lexeria commented on project/src/store/slices/data-slice/selector.test.ts:
```diff
@@ -81,10 +81,11 @@ describe('data selector', () => {
-    expect(result).toBe(isSimilarFilmsLoading);
+  it('should return areSimilarFilmsLoading', () => {
+    const { areSimilarFilmsLoading } =
+      state[SliceName.Data];
```
Тут точно должен быть отступ? Как будто есть строчки подлиннее без него

GoveRaven replied:
странно. В VSCode всё нормально. Может я залил изменения с прошедшемся преттиром, когда задивал какие-то изменения
![image](https://github.com/user-attachments/assets/187b9d58-7fb8-4f0b-97b9-860974ff9119)

GoveRaven committed:
изменил экспорт/импорт browserHistory

GoveRaven committed:
вынес в переменную redirectToRoute

GoveRaven committed:
добавил отступ

GoveRaven committed:
добавил тип

GoveRaven committed:
добавил отступ

GoveRaven committed:
добавил тип

GoveRaven committed:
преттир прошёлся

GoveRaven committed:
заменил isSimilarFilmsLoading на areSimilarFilmsLoading

GoveRaven committed:
добавил утилз

GoveRaven committed:
отредактировал, чтобы нэйминг соответствовал стилю camelCase

GoveRaven committed:
заменил && на ||

GoveRaven committed:
authLogin и authLogout не меняют статус isAuthStatusChecked. В тестах это поле не проверяется

GoveRaven committed:
поместил функции с созданием моковых данных в один файл
Пару моковых данных оставил в коде как есть, т.к. они встречаются лишь раз + добавил их названиям приставку 'mock'

GoveRaven committed:
изменил путь импорта

GoveRaven committed:
добавил 2 функции и поменял расширение файла

GoveRaven committed:
добавил data-testid атрибуты

GoveRaven committed:
добавил тесты

GoveRaven committed:
добавил изменение состояние поля стейта после завершения загрузки

GoveRaven committed:
удалил тип, который больше не используется

Lexeria commented on project/src/pages/sign-in.test.tsx:
```diff
@@ -0,0 +1,51 @@
+  });
+
+  it('Компонент должен правильно отрендериться', () => {
+    const expectedHeaderText = 'signInHeader';
```
Это не текст, а айдишник. Лучше дать другое название для переменной

А еще где-то айдишники в `camelCase`, а где-то в `snake-case`, советую выбрать единый стиль и его придерживаться

GoveRaven replied:
переписал

Lexeria commented on project/src/pages/sign-in.test.tsx:
```diff
@@ -0,0 +1,51 @@
+import { SignIn } from './sign-in';
+import userEvent from '@testing-library/user-event';
+
+describe('Компонент: sign-in', () => {
```
Это хоть и компонент, но все же речь про страницу. Названия компонентов, кст, обычно пишутся как-то так: `SignIn`

Lexeria replied:
Можно еще добавить тест, где будет проверка на логин, что после нажатия на отправить с корректными логином и паролем все хорошо и происходит редирект. Аналогично вариант, что для невалидных данных ничего не работает

Lexeria commented on project/src/pages/sign-in.tsx:
```diff
@@ -129,6 +134,7 @@ export function SignIn(): JSX.Element {
                   id="user-email"
                   required
                   ref={emailInputRef}
+                  data-testid="emailAddress"
```
Аналогичное замечание про единый стиль нейминга

GoveRaven replied:
переименовал

Lexeria commented on project/src/utils/mockcreaters.tsx:
Забыл дефис в названии и мб очепятался в слове `creators`

GoveRaven replied:
исправил

Lexeria commented on project/src/utils/mockcreaters.tsx:
```diff
@@ -0,0 +1,98 @@
+  return <HistoryRouter history={memoryHistory}>{component}</HistoryRouter>;
+}
+
+export function createMockComponentwithStore(
```
`With` с большой буквы, кажется

GoveRaven replied:
переименовал

Lexeria commented:
Кажется, мы еще в процесс CI не вернули запуск тестов

Lexeria commented on project/src/components/review-form.tsx:
```diff
@@ -23,7 +23,7 @@ export function ReviewForm(): JSX.Element {
 
   return (
-    <div className="add-review">
+    <div className="add-review" >
```
А зачем здесь пробел?

GoveRaven replied:
случайно поставил

Lexeria commented on project/src/pages/add-review.test.tsx:
```diff
@@ -0,0 +1,110 @@
+import { AnyAction } from '@reduxjs/toolkit';
+import { MockStore } from '@jedmao/redux-mock-store';
+
+describe('Компонент: add-review', () => {
```
Скорее страничка AddReview, но это мелочи

Lexeria commented on project/src/pages/add-review.test.tsx:
```diff
@@ -0,0 +1,110 @@
+  const myMockStore = createMockStore();
+  let component: JSX.Element;
+  let store: MockStore<unknown, AnyAction>;
+  beforeEach(() => {
```
Можно отступ добавить, но тож не критично

GoveRaven replied:
добавил

GoveRaven committed:
добавил data-testid аттрибуты

GoveRaven committed:
добавил тип для mockStore

GoveRaven committed:
добавил моковый store

GoveRaven committed:
добавил тесты для страницы add-review

GoveRaven committed:
переписал data-testid в snake-case

GoveRaven committed:
переименовал файлы с моковыми функцияями

Lexeria commented on project/src/pages/add-review.test.tsx:
```diff
@@ -0,0 +1,110 @@
+
+    expect(actions).toEqual([
+      'data/fetchChosenFilm/pending',
+      'data/fetchChosenFilm/rejected',
+      'data/postComment/pending',
+      'data/postComment/rejected',
```
Кажется, rejected на фоне того, что сетевые запросы в тестах не симулируются и завершаются не успешно. Чтобы все было хорошо, можно попробовать поработать с запросами:
```js
mockAxiosAdapter.onPost(APIRoute.PostComment).reply(200, fakeServerReplay);
// click
// Возможно, можно без диспатча и по-другому строчку ниже обыграть, просто если там серия диспатчей будет, то можно так
await Promise.all([
  mockStore.dispatch(postComment(reviewInfo)),
]);
const actions = store.getActions().map(({ type }) => type);
expect(actions.at(-1)).toEqual(postComment.fulfilled.type);
```

GoveRaven committed:
переименовал функцию

Lexeria commented on project/src/utils/mock-creaters.tsx:
```diff
@@ -88,7 +88,8 @@ export function createMockComponentwithStore(
     ThunkDispatch<TState, typeof axios, Action>
   >(middleware);
-  const mockStore = mockStoreCreator(initialState);
+  const mockStore: ReturnType<typeof mockStoreCreator> =
```
А точно нужно тип конкретизировать тут?

GoveRaven replied:
нет, не нужен

Lexeria commented on project/src/utils/mock-store.ts:
```diff
@@ -0,0 +1,41 @@
+  const mockUserSlice = createMockUserSlice();
+  const mockDataSlice = createMockDataSlice();
+  return {
+    USER: mockUserSlice,
+    DATA: mockDataSlice,
```
А у нас для названий слайсов не енамы используются?

GoveRaven replied:
да, енам

Lexeria commented on project/src/utils/mock-creaters.tsx:
Кажется очепятка в названии осталась — `creatOrs`

GoveRaven replied:
переименовал

GoveRaven committed:
убрал пробел

GoveRaven committed:
убрал тип

GoveRaven committed:
добавил отступ

GoveRaven committed:
переименовал файл

GoveRaven committed:
в наваниях слайсов используется enum

GoveRaven committed:
исправил значения testId в тестах

GoveRaven committed:
переместил HistoryRouter на компонент выше

GoveRaven committed:
убрал HistoryRouter и добавил HelmetProvider

GoveRaven committed:
добавил зависимости

GoveRaven committed:
добавил условие в useEffect

GoveRaven committed:
в функцию, создающую моковый компонент, добавил обёртку HelmetProvider

GoveRaven committed:
обновил тесты для страницы Add Review

GoveRaven committed:
удалил историю, т.к. она е используется

GoveRaven committed:
написал тесты для маршрутизации

GoveRaven committed:
вернул CI тесты

GoveRaven committed:
добавил отступы

GoveRaven committed:
добавил тесты CI

GoveRaven committed:
добавил IgnorePatterns

GoveRaven committed:
добавил зависимости

GoveRaven committed:
добавил зависимости

GoveRaven committed:
запускаем тесты вручную
